message/validation/peer_limiter.go

```diff
@@ -0,0 +1,101 @@
+package validation
+
+import (
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"golang.org/x/time/rate"
+)
+
+type RateLimiter struct {
+	rsaLimiter     *rate.Limiter
+	invalidLimiter *rate.Limiter
+	blockRequests  bool
+	lastBlocked    time.Time
+	mu             sync.Mutex // Mutex to protect blockRequests and lastBlocked
+}
+
+func NewRateLimiter(rsaLimit, invalidLimit rate.Limit) *RateLimiter {
+	return &RateLimiter{
+		rsaLimiter:     rate.NewLimiter(rsaLimit, int(rsaLimit)),
+		invalidLimiter: rate.NewLimiter(invalidLimit, int(invalidLimit)),
+	}
+}
+
+func (rl *RateLimiter) AllowRequest(blockingTime time.Duration) bool {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+
+	if rl.blockRequests {
+		if time.Since(rl.lastBlocked) > blockingTime {
+			rl.blockRequests = false
+		} else {
+			return false
+		}
+	}
+
+	if rl.rsaLimiter.Tokens() < 1.0 || rl.invalidLimiter.Tokens() < 1.0 {
+		rl.blockRequests = true
+		rl.lastBlocked = time.Now()
+		return false
+	}
+
+	return true
+}
+
+func (rl *RateLimiter) RegisterRSAError() {
+	rl.rsaLimiter.Allow()
+}
+
+func (rl *RateLimiter) RegisterInvalid() {
+	rl.invalidLimiter.Allow()
+}
+
+type PeerRateLimitManager struct {
```

Comment:
I think just PeerRateLimiter name can be sufficient, as a struct that handles RateLimiter instances for each peer